### PR TITLE
Spacecraft dao impl fix

### PR DESCRIPTION
**Pull request by Broseten**

### CODE REVIEW

**Comment by Broseten** [comment]
Build failing due to missing interaction with CraftComponent class...

File: spaceMissionManager-persistance/src/main/java/cz/muni/fi/Entity/Spacecraft.java [hunk 1]
```diff
@@ -1,10 +1,68 @@
...
+	@Id
 	private Long id;
 
+	private String type; //just probes or astronauts...
+
+	private String name;
+
+	@OneToMany
+	private List<CraftComponent> allComponents;
+
```
Comment (mlynarikj) [inline spaceMissionManager-persistance/src/main/java/cz/muni/fi/Entity/Spacecraft.java]: Dependency on mission is missing

File: spaceMissionManager-persistance/src/main/java/cz/muni/fi/Dao/SpacecraftDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+public class SpacecraftDaoImpl implements SpacecraftDao {
+
+    @PersistenceContext
+    private EntityManager em;
+
+    @Override
+    public void addSpacecraft(Spacecraft spacecraft) {
+        if (spacecraft == null){
+            throw new IllegalArgumentException("Spacecraft should not be null");
+        }
```
Comment (mlynarikj) [inline spaceMissionManager-persistance/src/main/java/cz/muni/fi/Dao/SpacecraftDaoImpl.java]: spacecraft should have at least 1 component, check whether id is not null,

File: spaceMissionManager-persistance/src/main/java/cz/muni/fi/Entity/Spacecraft.java [hunk 3]
```diff
@@ -1,10 +1,68 @@
...
+import java.util.List;
+
+@Entity
 public class Spacecraft {
 
-	private int unfinishedComponents;
-	private String name;
-	private String type;
+	//ATTRIBUTES
+	@Id
```
Comment (mlynarikj) [inline spaceMissionManager-persistance/src/main/java/cz/muni/fi/Entity/Spacecraft.java]: add generation strategy

File: spaceMissionManager-persistance/src/main/java/cz/muni/fi/Entity/Spacecraft.java [hunk 4]
```diff
@@ -1,10 +1,68 @@
...
 	private Long id;
 
+	private String type; //just probes or astronauts...
+
+	private String name;
```
Comment (mlynarikj) [inline spaceMissionManager-persistance/src/main/java/cz/muni/fi/Entity/Spacecraft.java]: make name unique and not nullable, add corresponding checks to Dao

File: spaceMissionManager-persistance/src/main/java/cz/muni/fi/Entity/Spacecraft.java [hunk 5]
```diff
@@ -1,10 +1,68 @@
...
+//		int count = allComponents.size();
+//		for (CraftComponent c : allComponents){
+//			if(c.isReadyToUse()){
+//				count--;
+//			}
+//		}
+//		return count;
+//	}
+
+
```
Comment (mlynarikj) [inline spaceMissionManager-persistance/src/main/java/cz/muni/fi/Entity/Spacecraft.java]: change equals and hash to be based on name of spacecraft

File: spaceMissionManager-persistance/src/main/java/cz/muni/fi/Dao/SpacecraftDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        em.persist(spacecraft);
+    }
+
+    @Override
+    public void removeSpacecraft(Spacecraft spacecraft) {
```
Comment (mlynarikj) [inline spaceMissionManager-persistance/src/main/java/cz/muni/fi/Dao/SpacecraftDaoImpl.java:37]: check if id is not null

File: spaceMissionManager-persistance/src/main/java/cz/muni/fi/Dao/SpacecraftDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Override
+    public List<Spacecraft> findAllSpacecrafts() {
+        return em.createQuery("select s from Spacecraft s", Spacecraft.class).getResultList();
+    }
+
+    @Override
+    public Spacecraft findSpacecraftById(Long id) {
+        return em.find(Spacecraft.class, id);
```
Comment (mlynarikj) [inline spaceMissionManager-persistance/src/main/java/cz/muni/fi/Dao/SpacecraftDaoImpl.java]: check if id is not null

File: spaceMissionManager-persistance/src/main/java/cz/muni/fi/Dao/SpacecraftDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Override
+    public void updateSpacecraft(Spacecraft spacecraft) {
```
Comment (mlynarikj) [inline spaceMissionManager-persistance/src/main/java/cz/muni/fi/Dao/SpacecraftDaoImpl.java:61]: same checks as create

File: spaceMissionManager-persistance/src/main/java/cz/muni/fi/Dao/SpacecraftDao.java [hunk 9]
```diff
@@ -22,12 +22,6 @@
...
 	/**
 	 * 
-	 * @param type
-	 */
-	List<Spacecraft> findAllSpacecrafts(String type);
-
-	/**
-	 * 
 	 * @param id
 	 */
```
Comment (mlynarikj) [inline spaceMissionManager-persistance/src/main/java/cz/muni/fi/Dao/SpacecraftDao.java]: add javadocs to all interfaces methods

**Review by mlynarikj — DISMISSED** [review]

File: spaceMissionManager-persistance/src/main/java/cz/muni/fi/Dao/SpacecraftDaoImpl.java [hunk 10]
```diff
@@ -21,6 +21,12 @@ public void addSpacecraft(Spacecraft spacecraft) {
         if (spacecraft == null){
             throw new IllegalArgumentException("Spacecraft should not be null");
         }
+        if(spacecraft.getId() != null){
+            throw new IllegalArgumentException("Spacecraft should be null - generated by the database");
+        }
+        if (spacecraft.getAllComponents().isEmpty()){
+            throw new IllegalArgumentException("Spacecraft should have at least one component");
+        }
         em.persist(spacecraft);
```
Comment (mlynarikj) [inline spaceMissionManager-persistance/src/main/java/cz/muni/fi/Dao/SpacecraftDaoImpl.java]: one more check - test if name of given entity is not null same in update

**Review by mlynarikj — CHANGES_REQUESTED** [review]
just one last thing

**Review by mlynarikj — APPROVED** [review]

**Review by jsmadis — APPROVED** [review]
